src/azul/service/responseobjects/hca_response_v5.py

```diff
@@ -233,32 +241,109 @@ def _construct_tsv_content(self):
+            self._iterate_hit(hit, sample_writer, participant_writer)
+
+        return self._create_zipped_bdbag({'participant.tsv': participant_tsv,
+                                          'sample.tsv': sample_tsv})
+
+    @staticmethod
+    def _create_zipped_bdbag(data: dict):
+        """Writes strings of file objects to disk, creates and returns BDBag.
+        :parameter data: keys are the file names, values the file objects
+        :returns bdbag: archived (zipped) BDBag with values of data as
+                        payloads
+        """
+        with TemporaryDirectory() as bag_path, \
+                TemporaryDirectory() as tsv_file_dir:
+            bag = bdbag_api.make_bag(bag_path)
+            data_path = os.path.join(bag_path, 'data')
+            for fname, fobj in data.items():
+                f = open(os.path.join(tsv_file_dir, fname), 'w')
+                f.write(fobj.getvalue())
+                f.close()
+
+            for tsv_file in data.keys():
+                copy(os.path.join(tsv_file_dir, tsv_file), data_path)
+            assert ['participant.tsv', 'sample.tsv'] == os.listdir(data_path)
+            bag = bdbag_api.make_bag(bag_path, update=True)  # update TSV checksums
+            assert bdbag_api.is_bag(bag_path)
+            bdbag_api.validate_bag(bag_path)
+            assert bdbag_api.check_payload_consistency(bag)
+
+            # TODO: in case of concurrency, is it one user = one lambda?
+            # Otherwise, there might be problems figuring out what TSV file
+            # belongs to which user...
+
+            return bdbag_api.archive_bag(bag_path, 'zip')
+
+    def _iterate_hit(self, es_search_hit, writer, participant_writer=None):
         hit_dict = es_search_hit.to_dict()
         assert len(hit_dict['contents']['files']) == 1
         file = hit_dict['contents']['files'][0]
-        file_fields = self._translate(file, 'contents.files')
-        for bundle in hit_dict['bundles']:
-            # FIXME: If a file is in multiple bundles, the manifest will list it twice. `hca dss download_manifest`
-            # would download the file twice (https://github.com/DataBiosphere/azul/issues/423).
-            writer.writerow(self._translate(bundle, 'bundles') + file_fields)
+        if self.format == 'tsv':
+            file_fields = self._translate(file, 'contents.files')
+
+            for bundle in hit_dict['bundles']:
+                # FIXME: If a file is in multiple bundles, the manifest will list it twice. `hca dss download_manifest`
+                # would download the file twice (https://github.com/DataBiosphere/azul/issues/423).
+                writer.writerow(
+                    self._translate(bundle, 'bundles') + file_fields)
+        else:  # construct manifest for BDBag
+            if participant_writer:
+                file_fields_participant = self._translate(file, 'contents.files')
+                participant_writer.writerow(file_fields_participant)
```

Comment:
This will create duplicates, won't it? I'd expect there to be one row per participant in this TSV. I may be wrong.

Reply:
Yes, that didn't eliminate duplicates if present. Fixed it.